zarr/v3/codecs/transpose.py

```diff
@@ -32,82 +31,70 @@ class TransposeCodecMetadata:
+                "There must not be duplicates in the `order` tuple. " + f"Got: {self.order}"
             )
-            assert all(0 <= x < array_metadata.ndim for x in configuration.order), (
+            assert all(0 <= x < ndim for x in self.order), (
                 "All entries in the `order` tuple must be between 0 and "
-                + f"the number of dimensions in the array. Got: {configuration.order}"
+                + f"the number of dimensions in the array. Got: {self.order}"
             )
-            order = tuple(configuration.order)
+            order = tuple(self.order)
 
-        return cls(
-            array_metadata=array_metadata,
-            order=order,
-        )
+        if order != self.order:
+            return evolve(self, order=order)
+        return self
 
     @classmethod
     def get_metadata_class(cls) -> Type[TransposeCodecMetadata]:
         return TransposeCodecMetadata
 
-    def resolve_metadata(self) -> CoreArrayMetadata:
-        from zarr.v3.metadata import CoreArrayMetadata
+    def resolve_metadata(self, chunk_metadata: ChunkMetadata) -> ChunkMetadata:
+        from zarr.v3.metadata import ChunkMetadata
 
-        return CoreArrayMetadata(
-            shape=tuple(
-                self.array_metadata.shape[self.order[i]] for i in range(self.array_metadata.ndim)
-            ),
+        return ChunkMetadata(
             chunk_shape=tuple(
-                self.array_metadata.chunk_shape[self.order[i]]
-                for i in range(self.array_metadata.ndim)
+                chunk_metadata.chunk_shape[self.order[i]] for i in range(chunk_metadata.ndim)
             ),
-            data_type=self.array_metadata.data_type,
-            fill_value=self.array_metadata.fill_value,
-            runtime_configuration=self.array_metadata.runtime_configuration,
+            data_type=chunk_metadata.data_type,
+            fill_value=chunk_metadata.fill_value,
+            runtime_configuration=chunk_metadata.runtime_configuration,
         )
 
-    async def decode(
-        self,
-        chunk_array: np.ndarray,
-    ) -> np.ndarray:
-        inverse_order = [0 for _ in range(self.array_metadata.ndim)]
+    async def decode(self, chunk_array: np.ndarray, chunk_metadata: ChunkMetadata) -> np.ndarray:
+        inverse_order = [0 for _ in range(chunk_metadata.ndim)]
         for x, i in enumerate(self.order):
             inverse_order[x] = i
         chunk_array = chunk_array.transpose(inverse_order)
         return chunk_array
 
     async def encode(
-        self,
-        chunk_array: np.ndarray,
+        self, chunk_array: np.ndarray, chunk_metadata: ChunkMetadata
     ) -> Optional[np.ndarray]:
         chunk_array = chunk_array.transpose(self.order)
         return chunk_array
 
-    def compute_encoded_size(self, input_byte_length: int) -> int:
+    def compute_encoded_size(self, input_byte_length: int, _chunk_metadata: ChunkMetadata) -> int:
```

Comment:
Thoughts on `compute_encoded_size` vs `encoded_size`, or `get_encoded_size`?

Reply:
All fine with me. I just copied it straight from the implementation recommendations in the spec.